scipy-1.0/ckdtree.tex

```diff
@@ -18,27 +18,20 @@
 in memory. Thus, the maximum heap size for this algorithm is conceptually determined
 as \texttt{np.arange(max(k))}.
 
-Also in 2015, SciPy added support for generating approximated sparse distance matrices 
-between \texttt{KDTree} objects (sets of points organized into a binary space 
-partitioning data structure\cite{Bentley:1975:MBS:361002.361007}). 
-The user provides a maximum distance value
-above which all distances between points in two provided \texttt{KDTree} objects
-are set to 0. The distance metric used between k-d trees is not constrained
-to the conventional L2 (Euclidean) norm---any Minkowski p-norm value
-bounded by 1 and infinity is valid. By default, a dictionary of keys
-based sparse matrix is the returned data structure. Previously-published
-results had suggested that a hashing approach to sparse matrix assembly
-is 7 times faster than constructing with compressed row format (CSR)\cite{10.1007/978-3-540-75755-9_107}.
-The dictionary object maps row and column indices to elements of the sparse matrix, which is
-efficient for matrix construction because all zero values are skipped in
-the dictionary mapping. The C++ level sparse matrix construction releases the Python
-global interpreter lock for increased performance. Once constructed, the
-dictionary of keys sparse matrix has the amortized constant time complexity 
-($O(1)$) for distance value retrieval that one would expect from a 
-dictionary\cite{Cormen:2001:IA:580470}. For efficiently performing arithmetic
-operations on the sparse matrix, SciPy allows the dictionary of keys
-sparse matrix to be directly converted to the common CSR, CSC, and COO
-data structures.
+In 2015, SciPy added support for generating approximated sparse distance matrices 
+between \texttt{KDTree} objects by zeroing all distances that exceed a user-provided value. 
```

Comment:
I haven't used this, so maybe it's right as written but...
I read "exceed" as "greater than". I would expect values _less than_ a threshold to be set to zero.

Reply:
I think it is fine as is -- effectively ignoring points beyond a certain distance probably narrows the search space between two kdtrees (two sets of points).

The docstring for `scipy.spatial.cKDTree.sparse_distance_matrix` indicates:

> Computes a distance matrix between two cKDTrees, leaving as zero any distance greater than `max_distance`.

Reply:
Can we try a numerical example?

Let's say `max_distance` is set as 1.
A distance is calculated to be 0.0001. It does not exceed `max_distance`, so it remains 0.0001.
Another distance is calculated to be 1.1. It exceeds `max_distance`, so this distance is converted to 0.

Do I understand correctly?

Reply:
Yes, that means that is not stored in the matrix and as @tylerjereddy said it speed up things (see [here](https://github.com/scipy/scipy/blob/v1.2.0/scipy/spatial/kdtree.py#L907)). I have changed to `ignoring`, better?

Reply:
I think, informally, dividing the points into the binary space partitioning tree means that when you have a distance cutoff you don't even have to bother calculating some of the distances because of the spatial relationships established by the tree leaves.

So you can just leave those ignored distances as 0 in the sparse matrix -- the algorithm gains efficiency and the data structure loses density.

Reply:
Expanding on that a little, if you divide an initial cube into many small cubes that represent leaves on a binary space partitioning tree, and you have a max distance that is no greater than say the width of two cubes, you don't need to look at points that are registered as distant neighbors on the tree (the cubes are known to be too far apart, just set distance to 0), but you do need to look at neighboring cubes & the points within them.

Reply:
Thanks. @tylerjereddy right, I just didn't know that 0 was just being used to mean "infinity" or "ignore". 
I've done this exact sort of thing but I was making a sparse "similarity" graph, where similarity is an inverse measure of distance.  For that it was a little more natural to ignore ("leave at zero") values smaller than a threshold.

Anyway, this is fine now.